lib/sqlgen.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+
+const { iter } = require('metarhia-common');
+const { escapeString } = require('./pg-utils.js');
+
+const allowedConditions = [
+  '=', '!=', '<', '<=', '>', '>=', 'like',
+];
+
+const supportedOps = {
+  'select': Set,
+  'selectDistinct': null,
+  'count': Array,
+  'avg': Array,
+  'min': Array,
+  'max': Array,
+  'sum': Array,
+  'where': Array,
+  'groupBy': Set,
+  'orderBy': Set,
+  'from': null,
+  'limit': null,
+  'offset': null,
+};
+
+const functionHandlers = {
+  'count': op => `count(${op.field})`,
+  'avg': op => `avg(${op.field})`,
+  'min': op => `min(${op.field})`,
+  'max': op => `max(${op.field})`,
+  'sum': op => `sum(${op.field})`,
+};
+
+const escapeValue = value => {
+  const type = typeof value;
+  if (type === 'number') return value;
+  if (type === 'string') return escapeString(value);
+  if (value instanceof Date) return `'${value.toISOString()}'`;
+  throw new TypeError('Unsupported value (${value}) type');
+};
+
+const checkType = (value, name, type) => {
+  if (typeof value !== type) {
+    throw new TypeError(
+      `Invalid '${name}' value (${value}) type, expected '${type}'`
+    );
+  }
+};
+
+class SelectBuilder {
```

Comment:
How can I do this if some of the `PostgresCursor`/`Cursor` methods are supported only in `MemoryCursor` (`map`, `filter` etc) and some other methods present here (`whereNot`, `wheheIn`) are not supported by the `Cursor` interface at all yet (I wanted to have them in the future)? Therefore I can only map `jsql -> sql` at the `.fetch` time or I'll have to store the flag of `canUseSQL` to determine if the user haven't called any of the 'unsupported' (`map` etc) methods and based on that either call a `selectBuilder.method` or just store it in the array of `this.jsql`. Furthermore I'll have to override *every* `Cursor` method in order to set this `canUseSQL` flag as I cannot otherwise determine if it is still possible to use `SQL` or I have to fallback to JS operations.

Reply:
It is better to collect operations in array or struct and convert them to sql when `.fetch` called. If unsupported operations found, they can be separated and passed to `MemoryCursor`. It's unbelievable that we will use another RDBMS, but even if we do, we'll create parent class `SQLCursor` to be ancestor for all SQL-based cursors.